### PR TITLE
[ntuple] Add initial in-memory index prototype

### PR DESCRIPTION
**Pull request by enirolf**
This PR adds (a first version of) the `RNTupleIndex`, which is an in-memory structure that maps RNTuple field values (or combinations thereof) to an entry index in the RNTuple for which the index was built. Currently, the index only resides in memory and thus has to be (re)build each time. `RNTupleIndex` will be used by the `RNTupleProcessor` to enable dataset joins and will be as transparent as possible to users. Currently, no public interface is foreseen.

At this point, no persistification is foreseen, but this might be added in the future. The implementation of the `RNTupleIndex` in this PR is hash-based. An implementation that is vector-based (but with the same interface) will also be considered.

The idea is to benchmark and evaluate both implementations (and potentially more). Based on the results we can decide which one to actually use (or alternatively make multiple implementations available if they show clear tradeoffs in different use cases).

### CODE REVIEW

File: tree/ntuple/v7/inc/ROOT/RField.hxx
Comment (silverweed) [inline tree/ntuple/v7/inc/ROOT/RField.hxx:547]: Are both the argument and the method intentionally not `const` ?

File: tree/ntuple/v7/inc/ROOT/RNTupleIndex.hxx
Comment (silverweed) [inline tree/ntuple/v7/inc/ROOT/RNTupleIndex.hxx:51]: This `std::vector` should probably be either a const ref or a `std::span` to avoid deep-copying it.

File: tree/ntuple/v7/inc/ROOT/RNTupleIndex.hxx
Comment (silverweed) [inline tree/ntuple/v7/inc/ROOT/RNTupleIndex.hxx:64]: This `std::vector` should probably be either a const ref or a `std::span`.

File: tree/ntuple/v7/inc/ROOT/RNTupleIndex.hxx
Comment (silverweed) [inline tree/ntuple/v7/inc/ROOT/RNTupleIndex.hxx:74]: This `std::vector` should probably be either a const ref or a `std::span`.

File: tree/ntuple/v7/inc/ROOT/RNTupleIndex.hxx
Comment (silverweed) [inline tree/ntuple/v7/inc/ROOT/RNTupleIndex.hxx:85]: This `std::vector` should probably be either a const ref or a `std::span`.

File: tree/ntuple/v7/inc/ROOT/RNTupleIndex.hxx
Comment (silverweed) [inline tree/ntuple/v7/inc/ROOT/RNTupleIndex.hxx:108]: This `std::vector` should probably be either a const ref or a `std::span`.

File: tree/ntuple/v7/inc/ROOT/RNTupleIndex.hxx
Comment (silverweed) [inline tree/ntuple/v7/inc/ROOT/RNTupleIndex.hxx:120]: Consider adding `valuePtrs.reserve(sizeof...(Ts));` here

File: tree/ntuple/v7/src/RNTupleIndex.cxx
Comment (silverweed) [inline tree/ntuple/v7/src/RNTupleIndex.cxx:28]: consider adding `reserve()` here

File: tree/ntuple/v7/src/RNTupleIndex.cxx
Comment (silverweed) [inline tree/ntuple/v7/src/RNTupleIndex.cxx:22]: consider adding `reserve()` here

File: tree/ntuple/v7/src/RNTupleIndex.cxx
Comment (silverweed) [inline tree/ntuple/v7/src/RNTupleIndex.cxx:78]: consider adding `fields.reserve()` here

**Review by silverweed — CHANGES_REQUESTED** [review]

**Review by jblomer — CHANGES_REQUESTED** [review]
Cool new functionality!

Some comments:
- I think that is a fundamental new functionality that should have a section in the architecture.md
- Instead of adding the `GetHash()` member function, perhaps we can use a "hash visitor". That would allow for using only available public interfaces. The `RPrintValueVisitor` can serve as a blueprint (the actual hash visitor class is probably better placed in `RNTupleIndex.[hxx,cxx]` or in a separate compilation unit).

File: tree/ntuple/v7/inc/ROOT/RNTupleIndex.hxx
Comment (vepadulano) [inline tree/ntuple/v7/inc/ROOT/RNTupleIndex.hxx:71]: Add destructor for rule of five

File: tree/ntuple/v7/inc/ROOT/RNTupleIndex.hxx
Comment (vepadulano) [inline tree/ntuple/v7/inc/ROOT/RNTupleIndex.hxx:79]: Comment is outdated

File: tree/ntuple/v7/src/RNTupleIndex.cxx
Comment (vepadulano) [reply tree/ntuple/v7/src/RNTupleIndex.cxx:19]: Yep, I agree

File: tree/ntuple/v7/test/ntuple_index.cxx
Comment (vepadulano) [inline tree/ntuple/v7/test/ntuple_index.cxx:66]: This pattern of creating a page source, then the index, then using the RNTupleReader seems a bit odd. For now the API is not the main concern since anyway we'll integrate it in the RDF processing. But something to consider in general

**Review by vepadulano — COMMENTED** [review]
I really like the direction of this new feature, it shows a great deal of thinking! I left just some small comments that came up while reading the code.

File: tree/ntuple/v7/src/RNTupleIndex.cxx
Comment (silverweed) [inline tree/ntuple/v7/src/RNTupleIndex.cxx:69]: Given that here we're not constructing the returned vector, but rather copying an existing one, we might want to return a `const vector<> *` instead.
But this depends on 3 factors which I'm not aware of, so I'll let you decide what's best:
1. how big can this vector get? If it's just a few values, the copy is no problem at all.
2. how hot will this method be?
3. if later we will want to change the implementation so that we actually need to construct a vector rather than just referencing it, how big of a problem would that be?

File: tree/ntuple/v7/test/ntuple_index.cxx
Comment (enirolf) [reply tree/ntuple/v7/test/ntuple_index.cxx:66]: This is only for testing. Initially, I added a `CreateIndex` method to the `RNTupleReader`, but since we want to use the index only internally (at least for the foreseeable future) I removed it.

File: tree/ntuple/v7/src/RNTupleIndex.cxx
Comment (enirolf) [reply tree/ntuple/v7/src/RNTupleIndex.cxx:69]: What would be the reason to return a const pointer rather than a const ref in this case?

> how big can this vector get? If it's just a few values, the copy is no problem at all.

Very likely it will only contain one element. Maybe in some exotic cases a couple of them (depending on pileup), but that will definitely not be a common case.

> how hot will this method be?
Since it's called by `GetFirstEntryNumber` it will be quite hot (i.e. once every event in the event loop). Possibly `GetFirstEntryNumber` can be refactored to not use this one.

> if later we will want to change the implementation so that we actually need to construct a vector rather than just referencing it, how big of a problem would that be?
I think not a problem at all, based the points you raised I think it makes sense to change it to a reference.

File: tree/ntuple/v7/inc/ROOT/RNTupleIndex.hxx
Comment (hahnjo) [reply tree/ntuple/v7/inc/ROOT/RNTupleIndex.hxx]: Coming back to this: I think `RIndexValue` now stores the hash of all field values. This will reduce the probability of collisions, but I think it doesn't solve the fundamental problem of finding the "right" entry in case of (a) collision(s). Based on my limited understanding, I think I see two alternatives:
1. Store the values themselves, as Jakob suggests. Then the `std::unordered_map` can (internally) compare the actual values in case of collisions.
2. Alternatively, the `RNTupleIndex` could accept the (rare) occurrence of collisions and use the `std::vector` as candidate entries where either the index or the call site have to (re)load the index column and check if it's actually the value you were looking for.

**Comment by hahnjo** [comment]
While thinking about collisions and the index storage, one thought that crossed my mind is to template the index based on the index field type(s). I'm not sure if that's a good idea, one of the immediate question being "do we want `RNTupleIndex<std::string>` and `RNTupleIndex<std::uint64_t>` to be different types?" But still maybe something to consider, it might simplify the field value storage (if needed) and the entire hashing business...

**Comment by enirolf** [comment]
> While thinking about collisions and the index storage, one thought that crossed my mind is to template the index based on the index field type(s). I'm not sure if that's a good idea, one of the immediate question being "do we want `RNTupleIndex<std::string>` and `RNTupleIndex<std::uint64_t>` to be different types?" But still maybe something to consider, it might simplify the field value storage (if needed) and the entire hashing business...

This is how I initially implemented it. Indeed it makes the index itself much more straightforward. However, when I started prototyping the actual join/unaligned friends it really didn't work without making that interface overly complicated so in the end I opted for a non-templated version. Perhaps there's some template trickery to still make it play nice with the foreseen interface (or allow for a slightly different but still simple enough interface), I will think about it for a bit.

File: tree/ntuple/v7/inc/ROOT/RNTupleIndex.hxx
Comment (enirolf) [reply tree/ntuple/v7/inc/ROOT/RNTupleIndex.hxx]: Ah, I realize now I may have misunderstood the original comment. It's a bit tricky since the field values can be of different types, but as you note in your other comment this could be addressed with templating. Perhaps this is indeed something to reconsider...

**Comment by hahnjo** [comment]
Not sure if it applies here, but a common "template trickery" is to have a type-punned base class that provides enough functionality to cater for the hot paths during the event loop. We already use this in RNTuple, for example with `RFieldBase` that can be conveniently passed around if you only need its interface functions.

**Comment by enirolf** [comment]
After going back and forth between different resolutions, I've decided to for now only allow integral-type fields as index fields (thus dropping floating-point types and strings), and storing their values as `std::uint64_t`. For now, we don't have any concrete examples where a non-integral field is desired to be used for indexing, so to stay within the scope of "initial prototype", I think this approach should suffice. 

While it will solve the potential index collisions, storing the separate values will incur some memory overhead, but preliminary benchmarks indicate it should be manageable (to be followed-up, though).

File: tree/ntuple/v7/src/RNTupleIndex.cxx
Comment (jblomer) [inline tree/ntuple/v7/src/RNTupleIndex.cxx:79]: I think it would be nice to make this method idempotent, i.e. start with
```
if (fIsBuilt)
   return;
```

File: tree/ntuple/v7/src/RNTupleIndex.cxx
Comment (jblomer) [inline tree/ntuple/v7/src/RNTupleIndex.cxx:85]: Perhaps better to explicitly enumerate the supported types, e.g. check `field->GetTypeName()` against a static `std::unordered_set<std::string>`. Or use a dynamic cast with the recent `RIntegralField`.

**Review by jblomer — APPROVED** [review]
Nice and clean! Some comments for consideration before merging.

**Comment by pcanal** [comment]
>  I've decided to for now only allow integral-type fields as index fields (thus dropping floating-point types 

Good.  Having a floating point 'as is' as the index would be a disaster due to the instability of equality.  (One could envision having a floating pointer index only the user provide a transformation from floating point to integer).

**Review by silverweed — APPROVED** [review]

**Pull request merged**
